File: quotebot-go/internal/blog_auction_task.go
package internal

import (
	"math"
	"quotebot/lib"
	"time"
)

// BlogAuctionTask handles pricing and publishing of blog quotes
type BlogAuctionTask struct {
	marketDataRetriever *lib.MarketStudyVendor
}

// NewBlogAuctionTask creates a new BlogAuctionTask instance
func NewBlogAuctionTask() *BlogAuctionTask {
	return &BlogAuctionTask{
		marketDataRetriever: lib.NewMarketStudyVendor(),
	}
}

// PriceAndPublish calculates the price and publishes the quote
func (b *BlogAuctionTask) PriceAndPublish(blog string, mode string) {
	avgPrice := b.marketDataRetriever.AveragePrice(blog)

	// FIXME should actually be +2 not +1

	proposal := avgPrice + 1
	timeFactor := 1

	if mode == "SLOW" {
		timeFactor = 2
	}

	if mode == "MEDIUM" {
		timeFactor = 4
	}

	if mode == "FAST" {
		timeFactor = 8
	}

	if mode == "ULTRAFAST" {
		timeFactor = 13
	}

	if int(proposal)%2 == 0 {
		proposal = 3.14 * proposal
	} else {
		now := time.Now()
		yearStart := time.Date(now.Year(), 1, 0, 0, 0, 0, 0, now.Location())
		dayOfYear := now.Sub(yearStart).Milliseconds() / 10000000
		proposal = 3.15 * float64(timeFactor) * math.Round(float64(dayOfYear))
	}

	lib.Publish(proposal)
}

File: quotebot-elixir/lib/blog_auction_task.ex
defmodule Quotebot.BlogAuctionTask do
  @moduledoc """
  BlogAuctionTask handles pricing and publishing of blog quotes.
  Uses hard-coded dependencies and time-based calculations.
  """

  defstruct [:market_data_retriever]

  def new do
    %__MODULE__{
      market_data_retriever: Quotebot.Thirdparty.MarketStudyVendor.new()
    }
  end

  def price_and_publish(task, blog, mode) do
    avg_price = Quotebot.Thirdparty.MarketStudyVendor.average_price(task.market_data_retriever, blog)

    # FIXME should actually be +2 not +1
    proposal = avg_price + 1

    time_factor =
      case mode do
        "SLOW" -> 2
        "MEDIUM" -> 4
        "FAST" -> 8
        "ULTRAFAST" -> 13
        _ -> 1
      end

    proposal =
      if rem(trunc(proposal), 2) == 0 do
        3.14 * proposal
      else
        now = DateTime.utc_now()
        day_of_year = Date.day_of_year(now)
        3.15 * time_factor * Float.round(day_of_year / 1.0)
      end

    Quotebot.Thirdparty.QuotePublisher.publish(proposal)
  end
end
